spikes/cluster-api/README.md

```diff
@@ -0,0 +1,405 @@
+
+![Bootstrap and pivot diagram](../../docs/diagrams/out/ClusterAPI-Bootstrap-Pivot.svg)
+
+## Helpful links
+
+- Intro to Cluster API <https://cluster-api.sigs.k8s.io/introduction.html>
+- Cluster API quickstart <https://cluster-api.sigs.k8s.io/user/quick-start.html>
+- Bootstrap and pivot process <https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html?#bootstrap--pivot>
+- Azure provider <https://capz.sigs.k8s.io/>
+- AKS support <https://capz.sigs.k8s.io/topics/managedcluster.html>
+- Glossary <https://cluster-api.sigs.k8s.io/reference/glossary.html>
+
+## Prerequisites
+
+- AZ CLI [download](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+- Docker [download](https://docs.docker.com/install/)
+- kubectl [download](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/)
+- clusterctl [download](https://cluster-api.sigs.k8s.io/user/quick-start.html#install-clusterctl)
+- Access to existing Kubernetes cluster or ability to create a local one
+  - example: k3d, kind, minikube, etc
+
+## Using Codespaces
+
+GitHub Codespaces is an easy way to get up and running with all the installed tools.
+
+To open with codespaces:
+
+- Click the Code button on this repository
+- Click the Codespaces tab
+- Click Create button
+
+## Bootstrap
+
+One of the desired outcomes of this spike is to use Cluster API to create the management cluster, leveraging the automation in the Cluster API infrastructure provider. But, Cluster API needs to run in a Kubernetes cluster. When using this repo's codespaces, K3d is used as the temoprary [bootstrap cluster](https://cluster-api.sigs.k8s.io/reference/glossary.html#bootstrap) that is responsible for creating the management cluster.
+
+### Create local bootstrap cluster
+
+Create k3d cluster.
+
+```bash
+
+mkdir -p ./spikes/cluster-api/capi-configs
+
+# see Makefile for more info
+make create
+
+```
+
+Verify cluster creation.
+
+```bash
+
+kubectl cluster-info
+kubectl get pods -A
+
+```
+
+### Gather configuration values for bootstrap cluster setup
+
+Setup the service principal details for Cluster API to use.
+
+```bash
+
+az login --use-device-code
+
+# verify subscription
+az account show
+
+# create service principal
+export AZURE_SUBSCRIPTION_ID="<SubscriptionId>"
+export SP_NAME="<ServicePrincipalName>"
+
+az ad sp create-for-rbac \
+  --name $SP_NAME \
+  --role contributor \
+  --scopes="/subscriptions/${AZURE_SUBSCRIPTION_ID}"
+
+```
+
+Setup environment variables.
+
+```bash
+
+# Setup identity variables.
+export AZURE_TENANT_ID="<Tenant>"
```

Comment:
For tenant,client,and secret. Do you know if it is possible to get them from the `az ad` cli?

Reply:
Yep it is possible. I left it as is here since the process of getting a service principal and credentials is different depending on where these commands are run, user permissions, etc.